integration/integration.go

```diff
@@ -0,0 +1,244 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integration
+
+import (
+	"errors"
+	"flag"
+	"io/ioutil"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3db/bootstrap"
+	"github.com/m3db/m3db/interfaces/m3db"
+	"github.com/m3db/m3db/network/server/tchannelthrift/node"
+	"github.com/m3db/m3db/network/server/tchannelthrift/thrift/gen-go/rpc"
+	"github.com/m3db/m3db/services/m3dbnode/server"
+	"github.com/m3db/m3db/storage"
+	xclose "github.com/m3db/m3db/x/close"
+	xtime "github.com/m3db/m3db/x/time"
+	"github.com/stretchr/testify/require"
+
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+)
+
+// TODO(xichen): if this gets too unwieldy, we can consider making these
+// test options or loading configurations from a config file.
+const (
+	// channelName is the channel name.
+	channelName = "integration-test"
+
+	// readRequestTimeout is the read request timeout.
+	readRequestTimeout = 5 * time.Second
+
+	// writeRequestTimeout is the write request timeout.
+	writeRequestTimeout = 5 * time.Second
+
+	// serverStateChangeTimeout is the time we wait for a server to change its state.
+	serverStateChangeTimeout = 30 * time.Second
+)
+
+var (
+	httpClusterAddr        = flag.String("clusterhttpaddr", "0.0.0.0:9000", "Cluster HTTP server address")
+	tchannelClusterAddr    = flag.String("clustertchanneladdr", "0.0.0.0:9001", "Cluster TChannel server address")
+	httpNodeAddr           = flag.String("nodehttpaddr", "0.0.0.0:9002", "Node HTTP server address")
+	tchannelNodeAddr       = flag.String("nodetchanneladdr", "0.0.0.0:9003", "Node TChannel server address")
+	errServerStartTimedOut = errors.New("server took too long to start")
+	errServerStopTimedOut  = errors.New("server took too long to stop")
+)
+
+type dataMap map[string][]m3db.Datapoint
+
+func databaseOptions() m3db.DatabaseOptions {
+	var opts m3db.DatabaseOptions
+	opts = storage.NewDatabaseOptions().NewBootstrapFn(func() m3db.Bootstrap {
+		return bootstrap.NewNoOpBootstrapProcess(opts)
+	})
+	return opts
+}
+
+func setup() (m3db.DatabaseOptions, *time.Time, error) {
+	dbOpts := databaseOptions()
+
+	// set up now fn
+	now := time.Now().Truncate(dbOpts.GetBlockSize())
+	nowFn := func() time.Time { return now }
```

Comment:
This is really racy as other parts of the code modifies with just `*now = x`, think you need an RLock here and a `setNow(...)` and `getNow(...)` that wraps the locking